cp/cp_config.c

```diff
@@ -0,0 +1,172 @@
+	 * Looks like it may break */
+	watch_config_change("/etc/cp/config/app_config.cfg", config_change_cbk);
+}
+
+void 
+init_spgwc_dynamic_config(struct app_config *cfg )
+{
+	// Read the config file, parse it and fill passed cfg data structure 
+	const char *entry = NULL;
+	unsigned int num_dp_selection_rules = 0;
+    unsigned int index;
+    LIST_INIT(&cfg->dpList);
+
+	struct rte_cfgfile *file = rte_cfgfile_load(APP_CONFIG_FILE, 0);
+	if (NULL == file)
+	{
+		fprintf(stderr, "App config file is missing, ignore error...");
+		return;
+	}
+
+	entry = rte_cfgfile_get_entry(file, "GLOBAL", "NUM_DP_SELECTION_RULES");
+	{
+		fprintf(stderr, "NUM_DP_SELECTION_RULES missing from app_config.cfg file, abort parsing");
+		return;
+	}
+	num_dp_selection_rules = atoi(entry);
+
+	for(index = 0; index <num_dp_selection_rules; index++)
+	{
+		char sectionname[64] = {0};
+		struct dp_info *dpInfo = NULL;
+        dpInfo = (struct dp_info *)calloc(1, sizeof(struct dp_info));
+
+		snprintf(sectionname, sizeof(sectionname),
+				"DP_SELECTION_RULES_%u", index);
+		entry = rte_cfgfile_get_entry(file, sectionname, "DPID");
+		if (entry)
+        {
+			dpInfo->dpId = atoi(entry);
+		} 
+		entry = rte_cfgfile_get_entry(file, sectionname, "DPNAME");
+		if (entry)
+        {
+			strncpy(dpInfo->dpName, entry, DP_SITE_NAME_MAX);
+		} 		
+		entry = rte_cfgfile_get_entry(file, sectionname, "MCC");
+		if (entry)
+        {
+          // TODO : handle 2 digit mcc, mnc 
+          dpInfo->key.mcc_mnc.mcc_digit_1 = (unsigned char )entry[0];
+          dpInfo->key.mcc_mnc.mcc_digit_2 = (unsigned char )entry[1];
+          dpInfo->key.mcc_mnc.mcc_digit_3 = (unsigned char )entry[2];
+		} 		
+		entry = rte_cfgfile_get_entry(file, sectionname, "MNC");
+		if (entry)
+        {
+          dpInfo->key.mcc_mnc.mnc_digit_1 = (unsigned char )entry[0];
+          dpInfo->key.mcc_mnc.mnc_digit_2 = (unsigned char )entry[1];
+          dpInfo->key.mcc_mnc.mnc_digit_3 = (unsigned char )entry[2];
+		} 
+		entry = rte_cfgfile_get_entry(file, sectionname, "ENB_ADDRESS");
+		if (entry)
+        {
+          	if (!inet_aton(entry, &dpInfo->key.enb_addrs[0]))
+			{       
+             // do nothing 
+			}
+		} 
+        LIST_INSERT_HEAD(&cfg->dpList, dpInfo, dpentries);
+	}
+
+	return;
+} 
+
+/* Given key find the DP. Once DP is found then return its dpId */
+uint64_t 
+select_dp_for_key(struct dp_key *key)
+{
+  fprintf(stderr, "Key - MCC = %d%d%d MNC %d%d%d TAC = %d", key->mcc_mnc.mcc_digit_1, 
+         key->mcc_mnc.mcc_digit_2, key->mcc_mnc.mcc_digit_3, key->mcc_mnc.mnc_digit_1, 
+         key->mcc_mnc.mnc_digit_2, key->mcc_mnc.mnc_digit_3, key->tac);
+  
+  struct dp_info *np; 
+  LIST_FOREACH(np, &appl_config->dpList, dpentries)
+  {
+    if(bcmp((void *)(&np->key.mcc_mnc), (void *)(&key->mcc_mnc), 3) != 0)
+      continue;
+    if(np->key.tac != key->tac)
+      continue;
+    return np->dpId;
+  }
+  return DPN_ID; /* 0 is invalid DP */ 
+}
```

Comment:
Too many printfs.. but makes sense since it is still being developed and tested. Probably needs some cleaning up with regards to replacing absolute paths (with relative paths) as correctly mentioned in one of your comments.

Reply:
Yes. Will keep it for some time and once settled I will get rid of them.

Reply:
About absolute path Vs Relative Path. 
Things are becoming bit complex with K8s deployment and also standard deployment.
In case of K8s deployment we get config in the /etc/xxx path and then initial script pushes that config in the path which app reads... Need to find a better way so that config update feature works for both type of deployment seamlessly.

Reply:
I just checked the helm chart and it doesn't necessarily load config /etc/xx first and then copy it to /opt/xx. I'll fix the helm chart if that helps.

Reply:
Helm chart suggest that ConfigMap is mounted on /etc path and then script is executed to move the config to /opt/xxx path. Can you please correct me ?

Reply:
@hyunsun  I was seeing statefuleset file. omec/omec-control-plane/templates/statefulset-spgwc.yaml

I see following section:
        volumeMounts:
        - name: cp-script
          mountPath: /opt/cp/scripts/spgwc-run.sh
          subPath: spgwc-run.sh
        - name: cp-config
          mountPath: /etc/cp/config. <<< this is where all cfg files are available. Let me know if I am missing.